File: Tree/#501. Find Mode in Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int>res;
    int maxFreq = 0, curFreq = 0, cursor = INT_MIN;
    
    void inorder(TreeNode* root){
        if(!root) return;
        inorder(root->left);
        if(cursor == root->val) curFreq++;
        else curFreq = 1;
        
        if(curFreq > maxFreq){
            res.clear();
            maxFreq = curFreq;
            res.push_back(root->val);
        }
        else if(curFreq == maxFreq){
            res.push_back(root->val);
        }
        
        cursor = root->val;
        inorder(root->right);
    }
    vector<int> findMode(TreeNode* root) {
        inorder(root);
        return res;
    }
};

File: Tree/#501. Find Mode in Binary Search Tree.py
class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        # Inorder traversal
        self.res = []
        self.cnt = 0
        self.max_cnt = 0
        self.prev = None
        self.inorder(root)
        return self.res
        

    def inorder(self, root):
        if not root:
            return None

        self.inorder(root.left)
        self.cnt = 1 if root.val != self.prev else self.cnt + 1
        if self.cnt == self.max_cnt:
            self.res.append(root.val)
        elif self.cnt > self.max_cnt:
            self.res = [root.val]
            self.max_cnt = self.cnt
        self.prev = root.val
        self.inorder(root.right)